File: examples/cloud/supabase_example.py
#!/usr/bin/env python3
"""
@file supabase_example.py
@brief Python example usage of Supabase integration with Satox SDK
@copyright Copyright (c) 2025 Satoxcoin Core Developers
@license MIT License
"""

import json
import asyncio
from satox_bindings.database import DatabaseManager, DatabaseType

async def main():
    print("=== Satox SDK Supabase Integration Example (Python) ===")
    
    try:
        # Initialize database manager
        db_manager = DatabaseManager()
        
        # Load Supabase configuration
        with open('config/cloud/supabase_config.json', 'r') as f:
            config = json.load(f)
        
        # Connect to Supabase
        connection_id = await db_manager.connect(DatabaseType.SUPABASE, config)
        if not connection_id:
            print("Error: Failed to connect to Supabase")
            return
        
        print(f"Successfully connected to Supabase with ID: {connection_id}")
        
        # Example: Create a table
        create_table_query = """
            CREATE TABLE IF NOT EXISTS users (
                id SERIAL PRIMARY KEY,
                username VARCHAR(255) UNIQUE NOT NULL,
                email VARCHAR(255) UNIQUE NOT NULL,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )
        """
        
        create_result = await db_manager.execute_query(connection_id, create_table_query)
        if create_result.success:
            print("Table created successfully")
        else:
            print(f"Error creating table: {create_result.error}")
        
        # Example: Insert data
        insert_query = """
            INSERT INTO users (username, email) 
            VALUES ('john_doe', 'john@example.com')
            ON CONFLICT (username) DO NOTHING
        """
        
        insert_result = await db_manager.execute_query(connection_id, insert_query)
        if insert_result.success:
            print("Data inserted successfully")
        else:
            print(f"Error inserting data: {insert_result.error}")
        
        # Example: Query data
        select_query = "SELECT * FROM users WHERE username = 'john_doe'"
        select_result = await db_manager.execute_query(connection_id, select_query)
        if select_result.success:
            print("Query executed successfully")
            print(f"Result: {json.dumps(select_result.rows, indent=2)}")
        else:
            print(f"Error querying data: {select_result.error}")
        
        # Example: Transaction
        transaction_queries = [
            "INSERT INTO users (username, email) VALUES ('jane_doe', 'jane@example.com')",
            "UPDATE users SET email = 'jane.updated@example.com' WHERE username = 'jane_doe'"
        ]
        
        transaction_result = await db_manager.execute_transaction(connection_id, transaction_queries)
        if transaction_result.success:
            print("Transaction executed successfully")
        else:
            print(f"Error executing transaction: {transaction_result.error}")
        
        # Disconnect
        disconnect_result = await db_manager.disconnect(connection_id)
        if disconnect_result:
            print("Disconnected from Supabase")
        else:
            print("Error disconnecting")
        
    except Exception as e:
        print(f"Exception: {e}")
    
    print("=== Example completed ===")

if __name__ == "__main__":
    asyncio.run(main())

File: examples/cloud/supabase_example.ts
/**
 * @file supabase_example.ts
 * @brief TypeScript example usage of Supabase integration with Satox SDK
 * @copyright Copyright (c) 2025 Satoxcoin Core Developers
 * @license MIT License
 */

import { DatabaseManager, DatabaseType } from '../../bindings/typescript/src/database';
import * as fs from 'fs';

async function main() {
    console.log('=== Satox SDK Supabase Integration Example (TypeScript) ===');
    
    try {
        // Initialize database manager
        const dbManager = new DatabaseManager();
        
        // Load Supabase configuration
        const configData = fs.readFileSync('config/cloud/supabase_config.json', 'utf8');
        const config = JSON.parse(configData);
        
        // Connect to Supabase
        const connectionId = await dbManager.connect(DatabaseType.SUPABASE, config);
        if (!connectionId) {
            console.error('Error: Failed to connect to Supabase');
            return;
        }
        
        console.log(`Successfully connected to Supabase with ID: ${connectionId}`);
        
        // Example: Create a table
        const createTableQuery = `
            CREATE TABLE IF NOT EXISTS users (
                id SERIAL PRIMARY KEY,
                username VARCHAR(255) UNIQUE NOT NULL,
                email VARCHAR(255) UNIQUE NOT NULL,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )
        `;
        
        const createResult = await dbManager.executeQuery(connectionId, createTableQuery);
        if (createResult.success) {
            console.log('Table created successfully');
        } else {
            console.error('Error creating table:', createResult.error);
        }
        
        // Example: Insert data
        const insertQuery = `
            INSERT INTO users (username, email) 
            VALUES ('john_doe', 'john@example.com')
            ON CONFLICT (username) DO NOTHING
        `;
        
        const insertResult = await dbManager.executeQuery(connectionId, insertQuery);
        if (insertResult.success) {
            console.log('Data inserted successfully');
        } else {
            console.error('Error inserting data:', insertResult.error);
        }
        
        // Example: Query data
        const selectQuery = "SELECT * FROM users WHERE username = 'john_doe'";
        const selectResult = await dbManager.executeQuery(connectionId, selectQuery);
        if (selectResult.success) {
            console.log('Query executed successfully');
            console.log('Result:', JSON.stringify(selectResult.rows, null, 2));
        } else {
            console.error('Error querying data:', selectResult.error);
        }
        
        // Example: Transaction
        const transactionQueries = [
            "INSERT INTO users (username, email) VALUES ('jane_doe', 'jane@example.com')",
            "UPDATE users SET email = 'jane.updated@example.com' WHERE username = 'jane_doe'"
        ];
        
        const transactionResult = await dbManager.executeTransaction(connectionId, transactionQueries);
        if (transactionResult.success) {
            console.log('Transaction executed successfully');
        } else {
            console.error('Error executing transaction:', transactionResult.error);
        }
        
        // Disconnect
        const disconnectResult = await dbManager.disconnect(connectionId);
        if (disconnectResult) {
            console.log('Disconnected from Supabase');
        } else {
            console.error('Error disconnecting');
        }
        
    } catch (error) {
        console.error('Exception:', error);
    }
    
    console.log('=== Example completed ===');
}

main().catch(console.error);
